doc/RawSyscallsDesign.md

```diff
@@ -0,0 +1,196 @@
+but the code in the `RawSyscalls` implementation cannot be unit tested. To
+minimize the amount of non-unit-testable code, `RawSyscalls` is designed to be a
+minimal wrapper around `asm!`.
+
+## Major Design Considerations
+
+**Note: This file uses the same register naming convention as the Tock 2.0
+syscalls TRD. Registers `r0`-`r4` refer to ARM registers `r0`-`r4` and RISC-V
+registers `a0`-`a4`.**
+
+1. Most system calls only clobber `r0`-`r4`, while Yield calls have a far longer
+   clobber list. It would be inefficient to share an `asm!` call between Yield
+   and another system call.
+1. Exit does not return. This initially seems to be an important distinction,
+   but it is not. `Syscalls` can use `unreachable_unchecked!` to tell the
+   compiler it doesn't exit, and it should be obvious the return values are all
+   unused.
+1. The compiler is unable to optimize away unused arguments. For example,
+   Memop's "get process RAM start address" operation only needs `r0` set, while
+   Memop's "break" operation needs both `r0` and `r1` set. If our inline
+   assembly calls "get process RAM start address" but sets both `r0` and `r1`,
+   the compiler doesn't know that `r1` will be ignored so setting that register
+   will not be optimized away. Therefore we want to set the minimum number of
+   argument registers possible.
+1. The cost of specifying unused return registers is only that of unnecessarily
+   marking a register as clobbered. Explanation: After inlining, an unused
+   register is marked as "changed by the assembly" but can immediately be
+   re-used by the compiler, which is the same as a clobbered register. System
+   calls should generally be inlined -- and even if they aren't, the unused
+   return values will probably be passed in caller-saved registers (this is true
+   for the C ABI, so probably true for the Rust ABI), which are treated as
+   clobbered regardless.
+1. Command always needs to set all four registers, because otherwise we risk
+   leaking data to capsules (which are untrusted for confidentiality). Exit,
+   Memop, and Yield do not have this concern because they are implemented by the
+   core kernel (which is trusted).
+
+## Yield Method Breakdown
+
+`yield-wait` does not take any arguments, while `yield-no-wait` takes 1.
+Therefore, to avoid setting arguments unnecessarily, `yield-wait` and
+`yield-no-wait` should be distinct functions.
+
+## Remaining Method Breakdown
+
+After taking care of `yield`, we break down the remaining system calls by the
+number of arguments they take. Here is the breakdown (Exit and Memop are listed
+twice because they have sub-operations with varying number of arguments).
+
+* **Zero:**
+* **One:** Exit, Memop
+* **Two:** Exit, Memop
+* **Three:**
+* **Four:** Command, Read-Only Allow, Read-Write Allow, Subscribe
+
+Therefore we need three more methods: `one_arg_syscall`, `two_arg_syscall`,
+`four_arg_syscall`.
```

Comment:
Feel free to correct me, since I'm not suuuper familiar with the syscall ABI, but it feels like it would be simpler to just provide `syscall0` through `syscall4` and not think about it, and let the higher level layer fuss with how to map "real" syscalls to these functions (even if some go unused).

(At least, in every userland I've ever built this is how I've done it; I just want to know if there's a reason you're not doing this.)

Reply:
Because Yield needs to be separate (its clobbers are very different), you would end up with 10 system calls, most of which are unused. I'd rather not do that.